### PR TITLE
rpc: add ability to export/import descriptor files in `listdescriptors` and `importdescriptors`

### PR DESCRIPTION
**Pull request by w0xlt**
This PR adds a new parameter `save_to_file` to `listdescriptors` RPC, allowing the user to save the descriptors directly to a file and also adds a new parameter `desriptor_file` to `importdescriptors` RPC, allowing the user to import the descriptors directly from a file.

The first and second commits add the `options` parameter in a way that maintains compatibility with the previous `private` parameter.

With this PR, exporting / importing descriptors and creating watch-only wallet is as simple as:
```
$ bitcoin-cli -rpcwallet="cold_wallet"  listdescriptors "{\"save_to_file\": \"/home/user/Downloads/desc.json\"}"
$ bitcoin-cli -rpcwallet="watch_only_wallet"  importdescriptors "{\"descriptor_file\": \"/home/user/Downloads/desc.json\"}"
```

Without this PR, the user can manually create a file from the `listdescriptors` output or use `jq` to extract the `descriptor` field from `listdescriptors` but the idea here is to make things more automated and standardized.

And this also opens the way to add the functionality of importing descriptors from files in the GUI,

### CODE REVIEW

**Comment by kristapsk** [comment]
What's the point? Why can't you just redirect `bitcoin-cli` output to the file?

**Comment by w0xlt** [comment]
@kristapsk with #25757, perhaps the purpose of this PR has become clearer.

From the #25757 description:
With these PRs, exporting / importing descriptor and creating watch-only wallet is as simple as:

```
$ bitcoin-cli -rpcwallet="cold_wallet"  listdescriptors "{\"save_to_file\": \"/home/user/Downloads/desc.json\"}"
$ bitcoin-cli -rpcwallet="watch_only_wallet"  importdescriptors "{\"descriptor_file\": \"/home/user/Downloads/desc.json\"}"
```

Without these PRs, the user can manually create a file from the `listdescriptors` output or use `jq` to extract the descriptor field from `listdescriptors` but the idea here is to make things more automated and standardized.

And this also opens the way to add the functionality of importing descriptors from files in the GUI,

**Comment by aureleoules** [comment]
Concept ACK considering #25757 exists. I think you should add the other PR's commits to this one.
I believe this makes exporting and importing descriptors easier.

File: test/functional/wallet_listdescriptors.py [hunk 1]
```diff
@@ -73,6 +76,17 @@ def run_test(self):
...
+        assert_equal(wallet.listdescriptors({'private': True}), wallet.listdescriptors(True))
+
+        desc_file = os.path.join(node.datadir, 'desc.json')
+        assert_equal({
+            'wallet_name': 'w2',
+            'descriptor_file': desc_file
+            }, wallet.listdescriptors({'save_to_file': desc_file}))
+
+        with open(desc_file, encoding="utf8") as json_file:
+            assert_equal(expected['descriptors'], json.load(json_file))
```
Comment (aureleoules) [inline test/functional/wallet_listdescriptors.py:90]: maybe add a case when the file already exists?

**Review by aureleoules — COMMENTED** [review]
I was able to save the descriptor file with
`./src/bitcoin-cli -regtest -rpcwallet=test listdescriptors '{"save_to_file": "my_desc"}'`
But I feel like the syntax is a bit much, is there a way to simplify it?

In this case, it would be simplier to just do `./src/bitcoin-cli -regtest -rpcwallet=test listdescriptors > my_desc`

File: test/functional/wallet_listdescriptors.py (hunk 1 above)
Comment (w0xlt) [reply test/functional/wallet_listdescriptors.py]: Done in https://github.com/bitcoin/bitcoin/pull/25747/commits/c8840eb855bf9b5ec0b31d7b544442551dd7955f.

**Comment by w0xlt** [comment]
@aureleoules As suggested, I merged the other PR's commit to this one.

Note that `./src/bitcoin-cli -regtest -rpcwallet=test listdescriptors > my_desc` does not create a descriptor file, but a file whose content is the RPC response. You cannot directly import the content of this file into `importdescriptors`.

The `save_to_file` option writes in the file only the `descriptors` attribute of the response.

**Comment by S3RK** [comment]
Concept NACK. Using shell redirection is enough for CLI. Probably this change is better fitting with the GUI.

**Comment by luke-jr** [comment]
Concept NACK. I think we're trying to eliminate file-based RPC methods? At the very least, we shouldn't make RPC methods that do *both* file-based and non-file-based access...

**Comment by w0xlt** [comment]
This PR assumes that using descriptor files is a less error-prone and less manual way of creating watch-only wallets than the current way.

If it is not the case, what is the current workflow to create a watch-only wallet ?

**Comment by luke-jr** [comment]
For that use case, I would suggest a RPC method that makes a watch-only copy of a loaded wallet.

**Comment by achow101** [comment]
This PR does not seem to have conceptual support. Please leave a comment if you would like this to be reopened.

**Pull request closed without merging**